### DISCUSSION TITLE
Server-side LaTeX rendering

### DISCUSSION THREAD

**Discussion by MinimizationOrg** [quarto-dev/quarto-cli#5227] [Feature Requests]
Hello,

Thank you for your great work! The set of features provided by Quarto is very impressive! 

We are converting a math heavy website from Jekyll to Quarto, and are surprised to learn that Quarto doesn't support server-side math rendering.

Would it be possible to consider adding this feature? 

Both MathJax and KaTeX support server-side rendering for many years now, so hopefully this will not be too complicated to achieve. Another idea could be perhaps using pandoc to parse LaTeX directly.

Thank you in advance for your reply!

**Comment by mcanouil** [comment]
If MathJax/KaTeX can do that, then it should work within Quarto as long as you configure MathJax/KaTeX to use such feature.

Also note that the Math equation feature is from Pandoc as described in <https://quarto.org/docs/output-formats/html-basics.html#latex-equations> and from Pandoc documentation <https://pandoc.org/MANUAL.html#math-rendering-in-html>.

**Comment by cderv** [comment]
Server side math rendering is indeed not support by Quarto as Pandoc does not support it out of the box. 
This would be an interesting feature for Quarto indeed. 

I know of a solution in R that offers KaTeX server side rendering, which a post processor function for HTML document produced by Pandoc 
https://docs.ropensci.org/katex/reference/pandoc.html

I'll convert to a feature request to follow discussion and future work on this.

**Reply by mcanouil** [reply]
I remembered a blog post on the matter which was edited to mention KaTeX and just found it, see <https://masalmon.eu/2021/06/15/serverside-mathjax/>.
<img width="886" alt="image" src="https://user-images.githubusercontent.com/8896044/232803540-a4efafe3-29f8-4d27-8908-94762051fdd1.png">

**Reply by cderv** [reply]
Yes - remember than one about the experimentation by Maelle.

**Comment by MinimizationOrg** [comment]
Thanks @cderv and @mcanouil! Looking forward to the official implementation. 

There are some existing Pandoc filters that could do the job. For example, for MathJax the following seem to work well:
- https://github.com/shreevatsa/pandoc-mathjax-filter
- https://github.com/pandoc/lua-filters/tree/master/math2svg

By the way, is there a way to **globally disable math rendering** in Quarto altogether? We are opting for a server-side rendering solution and it seems to have some incompatibility with the existing one.

**Reply by cderv** [reply]
Those are nice filters ! Should be quite useful to make them available as quarto extension ! 
But they should work alreadty with Quarto. Did you try them ? 

> By the way, is there a way to globally disable math rendering in Quarto altogether?

I believe this should be possible by doing 
````
html-math-method: plain
````
This is a format option available: https://quarto.org/docs/reference/formats/html.html#format-options

**Reply by MinimizationOrg** [reply]
Thanks @cderv.

> Those are nice filters ! Should be quite useful to make them available as quarto extension ! But they should work alreadty with Quarto. Did you try them ?

Yes I had tested them before posting the message and they were working well with Quarto as filters.

> I believe this should be possible by doing
> 
> ```
> html-math-method: plain
> ```
> This is a format option available: https://quarto.org/docs/reference/formats/html.html#format-options

This doesn't work actually, it still converted the equations to Unicode. There should be some `html-math-method: None` option to disable completely math rendering and leave the equations intact. I ended up using
```
html-math-method:
      method: mathjax
      url: ""
```
as a hack. Before that, I tried downloading the pandoc template from https://github.com/quarto-dev/quarto-cli/tree/main/src/resources/formats/html/pandoc and removing the following code block from `template.html`:
```html
$if(math)$
$if(mathjax)$
  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
$endif$
  $math$
$endif$
```
but then I didn't know how to use that template for our current website (I couldn't find the information in the documentation despite having spent quite a lot of time looking for it).

**Reply by mcanouil** [reply]
> This doesn't work actually, it still converted the equations to Unicode. There should be some html-math-method: None option to disable completely math rendering and leave the equations intact. I ended up using

It does work, but Pandoc tries to be smart.
![image](https://user-images.githubusercontent.com/8896044/233111047-81318771-1171-45cb-875e-f7c14251a6c9.png)

> as a hack. Before that, I tried downloading the pandoc template from https://github.com/quarto-dev/quarto-cli/tree/main/src/resources/formats/html/pandoc and removing the following code block from template.html:
> ...
> but then I didn't know how to use that template for our current website (I couldn't find the information in the documentation despite having spent quite a lot of time looking for it).

You need to use the partials, see <https://quarto.org/docs/journals/templates.html#html-partials>.

**Reply by cderv** [reply]
Thanks @mcanouil ! 

> disable completely math rendering and leave the equations intact.

What does disable math mean for you ? What "equations intact" are ? 

With you hack, it seems you still need Pandoc to parse the math in the markdown but you don't want Mathjax to be included an process the math ? Is that right ? 

I thought you wanted to do some Lua thing and so still need the Math to be recognized in AST but no specific processing as you would interfer. Hence `plain` method. 

So I may have missed what you want in your output.

**Reply by MinimizationOrg** [reply]
> It does work, but Pandoc tries to be smart.

@mcanouil But this is exactly what I said above: "it still converted the equations to Unicode". My question is how to make Pandoc stop being smart in this case ;) 

@cderv Sorry for not being clear. What I want is that, if I type "$\alpha + \beta$" in my document, then the final output HTML should contain exactly "$\alpha + \beta$" (and not α + β) and it should also not contain any external scripts such as the MathJax JS. I hope it's clear now but please let me know if addition information is needed.

**Reply by MinimizationOrg** [reply]
@cderv @mcanouil Just to be clear, this has become a very minor issue because the MathJax hack above already works for me; so please feel free to ignore it.

**Reply by cderv** [reply]
> What I want is that, if I type "$\alpha + \beta$" in my document, then the final output HTML should contain exactly "$\alpha + \beta$" (and not α + β)

I think what you would need is to deactivate the math extension `-tex_math_dollars`
````markdown
---
title: 'No math parsing'
format: 
  html:
    from: markdown-tex_math_dollars
---

$x+1$
````

This will output 
```html
<p>$x+1$</p>
```
if this is what you want. No pandoc parsing at all, so this is considered as usual test. 

But I did not think this is what you wanted because your workarournd does not do that 

```
html-math-method:
      method: mathjax
      url: ""
```

will do the parsing in Pandoc (in the AST for lua processing for example), and write HTML output as expected by mathjax, but you will have no JS processing in your output because no url to load.
The HTML produced is 
````html
<p><span class="math inline">\(x+1\)</span></p>
````

this is different.

**Reply by MinimizationOrg** [reply]
@cderv Thanks!

> I think what you would need is to deactivate the math extension `-tex_math_dollars`

Unfortunately this only works for `$ $` but not for environments (such as `\begin{equation}\end{equation}`). Sorry I should have been more comprehensive in my previous comments. 

What I want is for Quarto to completely ignore math equations. "Hey Quarto, could you please pretend to see no math in all my documents?"

> The HTML produced is
> 
> ```
> <p><span class="math inline">\(x+1\)</span></p>
> ```
> 
> this is different.

Indeed, but this kind of outputs is still acceptable to me. On one hand, the "math inline" span container has absolutely no effect. On the other hand, the `\( \)` delimiters are similar too `$ $`, so it's fine.

**Reply by cderv** [reply]
> Unfortunately this only works for $ $ but not for environments (such as \begin{equation}\end{equation}). Sorry I should have been more comprehensive in my previous comments

I don't understand. If you deactivate mathjax with the plain value and deactivate the extension to keep the dollar syntax, the raw latex in HTML should be omitted as expected. There is no math parsing, Quarto or Pandoc is doing nothing with Math. 
Do you want to keep the raw latex in your HTML output ? Because this is not related with math. 

I wonder if you should use `html-math-method: plain` and then use a Lua filter to catch the [Math](https://pandoc.org/lua-filters.html#type-math), and the raw latex you need, and do what you want with it.

**Reply by MinimizationOrg** [reply]
> Do you want to keep the raw latex in your HTML output ? Because this is not related with math.

So I think this is what I misunderstood, I thought it had to do with math parsing. Yes I do want to keep the raw LaTeX in my HTML output, and I'm looking for a simple solution to achieve this.

Using the following settings: 

```html
format:
  html:
    html-math-method: plain
    from: markdown-tex_math_dollars
```

I obtained the following results:

<img width="960" alt="Screen Shot 2023-04-19 at 23 41 25" src="https://user-images.githubusercontent.com/129618842/233206471-8951f12e-ac95-428d-b739-a2aead5704b4.png">


There are two issues here: 

1. The equation inside `\begin{equation}\end{equation}` has disappeared.
2. The other equations have turned into other things ("2+b" has become a superscript). Note that the superscript conversion did not happen outside the math delimiters (`2^x` and `x^2` remain the same), which is precisely why I thought it had to do with math. Are you sure the superscript conversion is not related to math parsing?

**Reply by MinimizationOrg** [reply]
**Update:** In my new test, the superscript also happens for `a^2+b^2=c^2` without math delimiters, so indeed this has nothing to do with math parsing.

**Reply by cderv** [reply]
> The equation inside \begin{equation}\end{equation} has disappeared.

Yes your LaTeX part will be seen as raw latex, and this is not supposed to be in a HTML output. You can force Pandoc to include that in your output by using [a raw LaTeX block I believe.](https://pandoc.org/MANUAL.html#extension-raw_attribute) or wrapping the content in some HTML tag you need. 

> Note that the superscript conversion did not happen outside the math delimiters (2^x and x^2 remain the same)

This is because by deactivating math parsing, you are telling pandoc that this is usual markdown syntax, and `^2+b^`  will be seen a the Markdown superscript syntax (https://quarto.org/docs/authoring/markdown-basics.html#text-formatting), unless you espace this to avoid superscript processing. 

What you want is almost a custom reader and / or writer that would parse the document differently than usual markdown and output differently than default Pandoc. I really think you should look into custom lua filter (through Quarto extension) so that you let Pandoc parse Math, but you catch them in Lua to write yourself the output you want. 

Hope this clarifies what is happening in your tests.

**Reply by MinimizationOrg** [reply]
@cderv Thanks for the explanation, that totally makes sense! I have some ideas for the Lua filter that I'm going to try, will keep you updated! By the way, is it possible to apply a filter globally (i.e. to all documents)?

**Reply by cderv** [reply]
> By the way, is it possible to apply a filter globally (i.e. to all documents)?

Yes you need to set configuration in `_quarto.yml`

Have a look at Extension also https://quarto.org/docs/extensions/ - you can store and share filter very easily this way.

**Reply by MinimizationOrg** [reply]
@cderv Great, thanks! I will share my filter for server-side MathJax rendering (which is just an upgrade of https://github.com/pandoc/lua-filters/tree/master/math2svg to MathJax v3).

**Comment by mcanouil** [comment]
Closing in favour of https://github.com/quarto-dev/quarto-cli/issues/5229
